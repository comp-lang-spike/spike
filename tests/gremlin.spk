#!/usr/bin/env spike

printf(format, ...args) { stdout.printf(format, ...args); }

obj g;

foo(x) {
    g = x;
}

class Gremlin {
    obj anotherGizmo;
    
    self eachGizmoDo: aBlock {
        aBlock("hey");
        aBlock(42);
        aBlock(anotherGizmo);
        return aBlock("tomato");
    }
    
    anotherGizmo = aGizmo {
        anotherGizmo = aGizmo;
    }
    
    crash(crashBlock) {
        crashBlock(self);
    }
    
    red() { self.green(); }
    green() { self.blue(); }
    blue() { self.huh; }

    self doesNotUnderstand: aMessage {
        printf("I don't understand '%s'!\n",
               aMessage.selector.printString);
    }
}

zeroToTen() {
    obj r, i;
    r = Array.new(10);
    for (i = 0; i < 10; ++i)
        r[i] = i;
    return r;
}

main(argv) {
    foo(666);
    printf("The global var is %s.\n", g.printString);
    
    obj gremlin, result;
    gremlin = Gremlin.new();
    gremlin.anotherGizmo = zeroToTen();
    
    result = gremlin eachGizmoDo: [ :gizmo | printf("%s\n", gizmo.printString); gizmo ];
    printf("The result is %s.\n", result.printString);
    
    gremlin.crash([ :g | g.red() ]);
    
    return 0;
}
