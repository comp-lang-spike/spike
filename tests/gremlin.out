The global var is 666.
"hey"
42
{ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 }
"tomato"
The result is "tomato".
I don't understand '$huh'!
